### DISCUSSION TITLE
New route handler format

### DISCUSSION THREAD

**Discussion by mike14747** [vercel/next.js#46452] [Help]
Is this basically how the new route handlers in v13.2 are intended to be used?

```ts
// app/api/users/route.ts

import { NextRequest, NextResponse } from 'next/server';
import { getInfoForAllUsers, registerNewUser } from '../../../lib/api';

export async function GET() {
    const data = await getInfoForAllUsers().catch(error => {
        console.log(error);
        return new Response(null, { status: 500 });
    });
    return data ? NextResponse.json(data) : new Response(null, { status: 500 });
}

export async function POST(request: NextRequest) {
    try {
        const body = await request.json();
        if (!body.username || !body.password || !body.email) return new Response(null, { status: 400 });

        const result = await registerNewUser(body.username, body.password, body.email);
        return result?.code ? new Response(null, { status: result.code }) : new Response(null, { status: 500 });
    } catch (error) {
        console.log(error);
        return new Response(null, { status: 500 });
    }
}
```

Seems like the only way to return data is using NextResponse.json() and the only way to send a custom response status is with a new Response. Is that the case?

**Comment by icyJoseph** [comment]
I guess at the moment yeah, since this is but a use case of https://developer.mozilla.org/en-US/docs/Web/API/Response - however, perhaps, in future releases, NextReponse (which extends Web Api Response), could be further enhanced to handle server failure messages, and others.

I can imagine some people want to have full control as well, while others want to abstract these away...

**Comment by khuezy** [comment]
What about standalone next servers? For example, if you're running a standalone nextjs server on AWS Lambda, it takes in a `Request` object, which needs to be relative. If you give it an absolute URL to itself, it'll be an infinite loop.

**Comment by devietti** [comment]
I was wondering the same thing. I found that these work for me on v13.2.3:

```typescript
export async function POST(request: Request) {
   // return JSON with custom status
   return NextResponse.json( "invalid request format", { status: 400 } )
}

export async function PUT(request: Request) {
   // return other formats with custom status
    return new Response('invalid request format', { status: 400 })
}
```
